Painless: Fix bug for static method calls on interfaces

jdconrad opened the pull request:
We were not calling static interface methods correctly leading to errors.  Java 8 didn't care, but Java 9+ is stricter.  Example --

```java.lang.IncompatibleClassChangeError: Method java.util.Comparator.comparing(Ljava/util/function/Function;)Ljava/util/Comparator; must be InterfaceMethodref constant```

StackOverflow explanation -- (https://stackoverflow.com/questions/34360718/invokestatic-on-static-method-in-interface?utm_medium=organic&utm_source=google_rich_qa&utm_campaign=google_rich_qa)

elasticmachine commented:
Pinging @elastic/es-core-infra

nik9000 commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/Definition.java:
```diff
@@ -202,16 +203,27 @@ public MethodType getMethodType() {
             if (java.lang.reflect.Modifier.isStatic(modifiers)) {
-                writer.invokeStatic(type, method);
-            } else if (java.lang.reflect.Modifier.isInterface(owner.clazz.getModifiers())) {
+                // special case for interfaces where the interface function needs to be set to true
```
I think this'd be more clear if you said something like "invokeStatic assumes that it is not invoking a method on an interface."

jdconrad replied:
Agreed.  Cleaned up a bit.

nik9000 commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/FunctionRef.java:
```diff
@@ -66,6 +66,9 @@
     public final Type delegateType;
 
+    /** whether a call is made on a delegate interface */
+    public final int isDelegateInterface;
```
I wonder if it'd be worth making this a `boolean` for readability. I'm fine with it staying an integer if you like but I think it is worth a comment why it is an integer and that it can only be 0 or 1.

jdconrad replied:
Fixed and commented.

nik9000 commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/FunctionRef.java:
```diff
@@ -97,10 +100,13 @@ public FunctionRef(Class<?> expected, Method interfaceMethod, Method delegateMet
+            isDelegateInterface = 0;
         } else if (delegateMethod.augmentation != null) {
             delegateClassName = delegateMethod.augmentation.getName();
+            isDelegateInterface = delegateMethod.augmentation.isInterface() ? 1 : 0;
```
This'd only come up if the target augmentation method is defined on an interface, right? Maybe we should not allow that at all.

jdconrad replied:
Agreed, but I'd like to do that in a follow up PR as we discussed.

jdconrad commented:
@nik9000 Thanks for the fast turn around on this review!  Added a test for method references triggering an invokeStatic through lambdaBootstrap and fixed up a couple things based on your comments.

jdconrad commented:
Will commit this as soon as it passes CI
